src/power_grid_model/core/serialization.py

```diff
@@ -0,0 +1,319 @@
+        pgc.serializer_get_to_binary_buffer(self._serializer, int(use_compact_list), byref(raw_data), byref(size))
+        assert_no_error()
+
+        result = raw_data[: size.value]
+        if not isinstance(result, bytes):
+            raise PowerGridSerializationError("Invalid output data type")
+
+        return result
+
+    @abstractmethod
+    def dump(self, *args, **kwargs):
+        """
+        Dump the data in the recommended data format (e.g. str or bytes, depending on the serialization_type).
+
+        Args:
+            *: any serialization_type dependent args
+        """
+
+
+class _BytesSerializer(Serializer):
+    """
+    Base type for serialization to bytes
+    """
+
+    dump = Serializer.dump_bytes
+
+
+class _StringSerializer(Serializer):
+    """
+    Base type for serialization to str
+    """
+
+    dump = Serializer.dump_str
+
+
+class JsonDeserializer(Deserializer):  # pylint: disable=too-few-public-methods
+    """
+    JSON deserializer for the Power grid model
+    """
+
+    def __new__(cls, data: Union[str, bytes]):
+        return super().__new__(cls, data, SerializationType.JSON)
+
+
+class MsgpackDeserializer(Deserializer):  # pylint: disable=too-few-public-methods
+    """
+    msgpack deserializer for the Power grid model
+    """
+
+    def __new__(cls, data: bytes):
+        return super().__new__(cls, data, SerializationType.MSGPACK)
+
+
+class JsonSerializer(_StringSerializer):  # pylint: disable=too-few-public-methods
+    """
+    JSON deserializer for the Power grid model
+    """
+
+    def __new__(
+        cls,
+        data: Union[Mapping[str, np.ndarray], Mapping[str, Union[np.ndarray, Mapping[str, np.ndarray]]]],
+        dataset_type: Optional[str] = None,
+    ):
+        return super().__new__(cls, data, SerializationType.JSON, dataset_type=dataset_type)
+
+
+class MsgpackSerializer(_BytesSerializer):  # pylint: disable=too-few-public-methods
+    """
+    msgpack deserializer for the Power grid model
+    """
+
+    def __new__(
+        cls,
+        data: Union[Mapping[str, np.ndarray], Mapping[str, Union[np.ndarray, Mapping[str, np.ndarray]]]],
+        dataset_type: Optional[str] = None,
+    ):
+        return super().__new__(cls, data, SerializationType.MSGPACK, dataset_type=dataset_type)
+
+
+def json_deserialize(data: Union[str, bytes]) -> Dict[str, Union[np.ndarray, Dict[str, np.ndarray]]]:
+    """
+    Load serialized JSON data to a new dataset.
+
+    Args:
+        data: the data to deserialize.
+
+    Raises:
+        ValueError if the data is inconsistent with the rest of the dataset or a component is unknown.
+        PowerGridError if there was an internal error.
+
+    Returns:
+        A tuple containing:
+            the deserialized dataset in Power grid model input format.
+            the type of the dataset.
+    """
+    result = JsonDeserializer(data).load()
+    assert_no_error()
+    return result
+
+
+def json_serialize(
```

Comment:
TBD here and for `msgpack` and the `*_to_file` serialization functionality:

> > > > @TonyXiang8787 @petersalemink95 open question on forward-compatibility:
> > > do you think it's a good idea to add an optional `version: str = "1.0"` keyword argument to the new `json_serialize_*` and `msgpack_serialize_*` interfaces? that way, users can decide whether they want to hard fix their format version if they themselves use a very specific output query or that they allow the PGM to serialize using the most recent version
> > 
> > 
> > I don't think that's needed. In a given version of PGM, we only serialize to the latest supported format.
> 
> A valid remark that @petersalemink95 had is that some people may use the serialized data as their temporary format and that they could load into another application/pipeline step. in that case, people may already be using specific queries.
> 
> If ever, I do believe that now is a good time to add such a feature (probably it's better to use an enum now that i think about it), because we would want to bump the default value when we want to. It also would enable us to develop on `main` instead of on a feature branch like we're doing now to prevent conflicts.
> 
> let's discuss on Monday. I do believe it to be out of scope of this PR, though

Reply:
@mgovers after discussion with @petersalemink95 we have decided not to support this. This would mean maintaining two or multiple versions of serializer in the future which is not feasible.

What we need to guarantee is the forward/backward compatibility of serialization format. If we release version `1.1` of serialization format, the new `json/msgpack` file should be readable with the old PGM serializer, only with the new features missing.